src/transport_ofi.c

```diff
@@ -678,6 +678,21 @@ int allocate_recv_cntr_mr(void)
     /* Register separate data and heap segments using keys 0 and 1,
      * respectively.  In MR_BASIC_MODE, the keys are ignored and selected by
      * the provider. */
+#ifdef ENABLE_MR_ENDPOINT
```

Comment:
Hey Guys, sorry about lobbing the passing comment, just wanted to share a little historical context on this ifdef spaghetti.

There are a variety of things in the OFI transport that date back to the early days and no longer appropriate/correct given current libfabric. For example, this code should really be gated off of whether the provider returns `FI_MR_PROV_KEY` on the domain. We didn't have this back when Mr. Basic and Mr. Scalable were a thing. Enabling user-selected keys (`!FI_MR_PROV_KEY`) only for Mr. Endpoint is incorrect because a fabric can choose to return both `FI_MR_ENDPOINT` and `FI_MR_PROV_KEY`. Resolving the key value at runtime adds instructions to the critical path, so you probably want a build flag that will enable support for provider-selected memory registration keys.

Reply:
@jdinan Thanks for your input here. The OFI manpage says that for `FI_MR_ENDPOINT`, memory registration occurs at the endpoint level, rather than domain. In such case, does it matter if provider returns `FI_MR_PROV_KEY`? 

We are also looking at #1077 where MR flags will be selected based on the provider in user, thus limit the users to provide any option for the MR flags. As you said, there is no build flag for `FI_MR_PROV_KEY` from the user. Does it make sense to select these flags based on the provider rather provide the option to the user? 

Also, looking at the provider feature matrix, it looks like `FI_MR_PROV_KEY` is required in conjunction with `FI_MR_ALLOCATED` and `FI_MR_LOCAL` - almost equivalent to what we have in SOS for `--enable-ofi-mr=basic`. I am trying to understand, if having a separate build flag for `FI_MR_PROV_KEY` will lead to the same flags for basic.

Reply:
Thanks @jdinan - we definitely missed that.

The comment at line 679 suggests that these user-selected keys are ignored by the provider when in MR_BASIC (FI_MR_PROV_KEY)  mode, but I'm not seeing anything specifying that in the libfabric manpages.  I'm thinking we should be careful here and only provide keys when the provider does not require `FI_MR_PROV_KEY` and use `fi_mr_key` when it does.

Also, is resolving the key value here really on the critical path?  I thought this occurred in `shmem_transport_init -> shmem_transport_ofi_target_ep_init`.

Reply:
I realized later that this particular ifdef block for MR-ENDPOINT was to ensure that an `uint64_t` key is provided. I confirmed with OFI developers that we can use this key for all MR modes except `FI_MR_RAW`. So, I removed the specific check for MR-ENDPOINT. 
@jdinan - Does this look ok to you?

Reply:
SOS uses `ENABLE_MR_SCALABLE` as the build flag for determining whether to support `FI_MR_PROV_KEY`. For an example, you can look at `shmem_transport_ofi_get_mr` where we choose at compile time whether to reference the table of MR keys. This is on the critical path, so you may want to have a build flag if you want to continue to support a static decision regarding provider-selected MR keys. Apologies, I haven't looked at #1077 to see if this is addressed there. IIUC, `FI_MR_BASIC` and `FI_MR_SCALABLE` have been long deprecated and it would be a healthy change to remove them from all of the SOS MR logic.

Reply:
Thanks @jdinan. In #1077, we are choosing MR modes based on the provider is use through `FI_PROVIDER` or `SHMEM_OFI_PROVIDER`. So, all the checks for `shmem_transport_ofi_get_mr` will be done at runtime as there will be no build flags. We may need to rethink there if that is adding too much overhead.